internal/generate/generate.go

```diff
@@ -0,0 +1,26 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package generate
+
+// Generator is an implementation of the GeneratorSDK interface
+type Generator struct{}
+
+// GeneratorSDK is an interface that specifies the Generate methods
+// to generate and write various OperatorSDK types such as
+// package manifests
+type GeneratorSDK interface {
+	//GeneratePackageManifest(opts *PkgOptions) error
+	GeneratePackageManifest(operatorName, version, outputDir string, opts ...*PkgOptions) error
+}
```

Comment:
This interface setup seems to not follow reusable/extensible principles, since we'd have to add/remove methods whenever we add/remove a generator. I'd rather see something like
```go
type Generator interface {
	Generate(*Context) error
}

type Context struct {
	GetWriter func(string) (io.Writer, error)
	// Other fields used for writing
}
```

then we can do something like
```go
gen := NewPackageManifestGenerator(operatorName, version, outputDir string, opts ...*PkgOptions)
ctx := &Context{}
ctx.GetWriter = func(path string) {
	f, err := os.Open(path)
	return f, err
}
if err := gen.Generate(ctx); err != nil {
	return err
}
```

Reply:
So, the issue there is the context would have to have all the info for all of the generate methods - which I think would be confusing. I also really don't like the callback method structure used in the original interface, and would much prefer to just have simple arguments passing the data into the generate method. (Think about how this will be tested in the faked out interface in the cmd layer. How am I going to check that the callback methods are set correctly? With an argument I can just make a simple assertion that the args are the correct values).

While the implementation I went with isn't the most minimalist, is it really that bad? How often will we be adding/removing generators? It seems like it's a write-once keep-forever kind of deal. I did stick the stuff that's likely to change (the optional args) into a struct, so the interface wont have to be changed and the fake won't have to be regenerated when we make changes.

Reply:
What is this `Generator` interface is buying us? Wouldn't we always be creating and using the concrete type on the cmd side? Is there anywhere that we would actually use a generic `Generator` interface other than to write a high level test?

It's totally possible that I'm missing the context with this PR, but perhaps a more valuable way to test the generators would be to restructure the internals as interfaces and/or [overridable functions](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader.go#L42-L44) such that we could [substitute different mock pieces in and out](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader_test.go#L73-L75) to more easily exercise the real code paths?

IMO we should make the cmd side _extremely simple_, to the point that it wouldn't really be valuable to write tests for it.

And the actual implementation that's called by the cmd code is where we should refactor what we have into nice testable standalone libraries.

This is pretty much how I structured `kubectl-operator` (between [`internal/cmd`](https://github.com/operator-framework/kubectl-operator/tree/main/internal/cmd) and [`internal/pkg/action`](https://github.com/operator-framework/kubectl-operator/tree/main/internal/pkg/action)) though I haven't gotten around to writing tests for it yet.

Reply:
I understand that the tool generates PkgManifest and Bundle. Am I right? 

Then, besides I am not very familiarized with this piece of code since I do not work on it. Shows that the appropriated design pattern that could be adopted to solve its is [Builder](https://golangbyexample.com/builder-pattern-golang/).

Reply:
> restructure the internals as interfaces and/or [overridable functions](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader.go#L42-L44) such that we could [substitute different mock pieces in and out](https://github.com/operator-framework/operator-lib/blob/aac7eed1886158f2b42c11ade64b9e0673a5d3d6/leader/leader_test.go#L73-L75) to more easily exercise the real code paths

This is what I am attempting to do by this PR. I wasn't including the cmd part just yet to keep things simple - I will add it on top of this PR.

Reply:
Hi @jberkhahn Could we refractory it adopting a design pattern? In this case, I cannot see a better option than the [Builder](https://golangbyexample.com/builder-pattern-golang/). However, I am open to suggestions. 

I just think that we need to start looking for a better Maintainability.

Reply:
I don't think Builder is a good fit for this particular bit. The cmd layer calls out to the generate package to generate a thing and write it to disk or stdout - it doesn't really care about getting the object back or even how it's constructed in the first place. I'm hoping my new implementation is much more maintainable. My intent is when a new generator is added to the interface, the method has required args as args and then a struct containing optional params. That way, if we need to change something or add functionality later, you just add a new field on the opts struct and change the implementation/tests. The interface doesn't have to change and the fake doesn't have to be regenerated, so again the only thing that needs to change in the next layer up is the implementation/tests.

Reply:
I am maybe getting wrong but shows that your motivation here is to change the code for you be able to test it easily as you did for other commands and not to shape the solution base on what it does and represents. 

I think has a work in progress regards this which is related to add the `run bundle` and then, the next step after that would be cleanup/shape the code. 

So, your tests maybe would better fit after that. However, if @jmrodri @estroz which I think are working on with be ok with move forward with what you are proposing here and it will not affect the wip .. so I am fine with it as well.